File: app/src/main/java/pl/trikimusic/controller/core/bluetooth/ConnectionActivityLease.kt
package pl.trikimusic.controller.core.bluetooth

import kotlin.math.abs
import pl.trikimusic.controller.domain.model.FilteredSensorData

class ConnectionActivityLease(
    private val idleTimeoutNanos: Long = DEFAULT_IDLE_TIMEOUT_NANOS,
    private val accelerationDeltaThresholdG: Float = DEFAULT_ACCELERATION_DELTA_THRESHOLD_G,
    private val gyroscopeThresholdDps: Float = DEFAULT_GYROSCOPE_THRESHOLD_DPS,
) {
    private var lastActivityTimestampNanos: Long? = null
    private var parkingRequested = false

    init {
        require(idleTimeoutNanos > 0L)
        require(accelerationDeltaThresholdG > 0f && accelerationDeltaThresholdG.isFinite())
        require(gyroscopeThresholdDps > 0f && gyroscopeThresholdDps.isFinite())
    }

    fun observe(
        sample: FilteredSensorData,
        explicitActivity: Boolean,
    ): Boolean {
        val sensorActivity =
            abs(sample.accelerationMagnitude - 1f) >= accelerationDeltaThresholdG ||
                sample.gyroscopeMagnitude >= gyroscopeThresholdDps
        return observe(sample.source.timestampNanos, explicitActivity || sensorActivity)
    }

    fun observe(timestampNanos: Long, active: Boolean): Boolean {
        if (timestampNanos < 0L) {
            reset()
            return false
        }
        val previous = lastActivityTimestampNanos
        if (previous == null || timestampNanos < previous) {
            lastActivityTimestampNanos = timestampNanos
            parkingRequested = false
            return false
        }
        if (active) {
            lastActivityTimestampNanos = timestampNanos
            parkingRequested = false
            return false
        }
        if (!parkingRequested && timestampNanos - previous >= idleTimeoutNanos) {
            parkingRequested = true
            return true
        }
        return false
    }

    fun reset() {
        lastActivityTimestampNanos = null
        parkingRequested = false
    }

    companion object {
        const val DEFAULT_IDLE_TIMEOUT_NANOS = 12_000_000_000L
        const val DEFAULT_ACCELERATION_DELTA_THRESHOLD_G = 0.08f
        const val DEFAULT_GYROSCOPE_THRESHOLD_DPS = 5f
    }
}

File: windows/TrikiMusicController.Windows/Core/ConnectionActivityLease.cs
using TrikiMusicController_Windows.Models;

namespace TrikiMusicController_Windows.Core;

public sealed class ConnectionActivityLease
{
    public const long DefaultIdleTimeoutNanos = 12_000_000_000;
    public const float DefaultAccelerationDeltaThresholdG = 0.08f;
    public const float DefaultGyroscopeThresholdDps = 5f;

    private readonly long _idleTimeoutNanos;
    private readonly float _accelerationDeltaThresholdG;
    private readonly float _gyroscopeThresholdDps;
    private long? _lastActivityTimestampNanos;
    private bool _parkingRequested;

    public ConnectionActivityLease(
        long idleTimeoutNanos = DefaultIdleTimeoutNanos,
        float accelerationDeltaThresholdG = DefaultAccelerationDeltaThresholdG,
        float gyroscopeThresholdDps = DefaultGyroscopeThresholdDps)
    {
        if (idleTimeoutNanos <= 0) throw new ArgumentOutOfRangeException(nameof(idleTimeoutNanos));
        if (!float.IsFinite(accelerationDeltaThresholdG) || accelerationDeltaThresholdG <= 0)
            throw new ArgumentOutOfRangeException(nameof(accelerationDeltaThresholdG));
        if (!float.IsFinite(gyroscopeThresholdDps) || gyroscopeThresholdDps <= 0)
            throw new ArgumentOutOfRangeException(nameof(gyroscopeThresholdDps));
        _idleTimeoutNanos = idleTimeoutNanos;
        _accelerationDeltaThresholdG = accelerationDeltaThresholdG;
        _gyroscopeThresholdDps = gyroscopeThresholdDps;
    }

    public bool Observe(FilteredSensorData sample, bool explicitActivity)
    {
        var sensorActivity = Math.Abs(sample.AccelerationMagnitude - 1f) >= _accelerationDeltaThresholdG ||
            sample.GyroscopeMagnitude >= _gyroscopeThresholdDps;
        return Observe(sample.Source.TimestampNanos, explicitActivity || sensorActivity);
    }

    public bool Observe(long timestampNanos, bool active)
    {
        if (timestampNanos < 0)
        {
            Reset();
            return false;
        }
        if (_lastActivityTimestampNanos is not long previous || timestampNanos < previous)
        {
            _lastActivityTimestampNanos = timestampNanos;
            _parkingRequested = false;
            return false;
        }
        if (active)
        {
            _lastActivityTimestampNanos = timestampNanos;
            _parkingRequested = false;
            return false;
        }
        if (!_parkingRequested && timestampNanos - previous >= _idleTimeoutNanos)
        {
            _parkingRequested = true;
            return true;
        }
        return false;
    }

    public void Reset()
    {
        _lastActivityTimestampNanos = null;
        _parkingRequested = false;
    }
}
